Draft version of the changelog 5.0 (#6927)
edgedb/edgedb: docs/changelog/5_x.rst

====
v5.0
====

:edb-alt-title: EdgeDB v5

To play with the new features, install the CLI using `our installation guide

> <https://www.edgedb.com/install>`_ and initialize a new project.

Reviewer: For the duration of prerelease, we must tell them to install nightly.

I think it is plausible that it will get merged by friday. See https://github.com/edgedb/edgedb-cli/pull/1217

.. code-block:: bash

>   $ edgedb project init

Reviewer 2: ```suggestion
  $ edgedb project init --server-version 5.0-beta.1
```
or whatever the name of this beta ends up being. I assume this is it but maybe I'm wrong.

Upgrading
=========

**Local and Cloud instances**

To upgrade a local project, first ensure that your CLI is up to date with
``edgedb cli upgrade``. Then run the following command inside the project
directory.

.. code-block:: bash

  $ edgedb project upgrade

Alternatively, specify an instance name if you aren't using a project.

.. code-block:: bash

  $ edgedb instance upgrade -I my_instance

The CLI will first check to see if your schema will migrate cleanly to EdgeDB
5.0. If the upgrade check finds any problems, it will report them back to you.

**Hosted instances**

To upgrade a remote (hosted) instance, we recommend the following
dump-and-restore process.

> .. XXX: Did we bump our Postgres to 16.2 as per #6838?

Reviewer: Yes, we now bundle Postgres 16.2

1. EdgeDB v5.0 only supports PostgreSQL 14 (or above). So check the version of
   PostgreSQL you are using before upgrading EdgeDB. If you're using Postgres
   13 or below, you should upgrade Postgres first.

2. Spin up an empty 5.0 instance. You can use one of our :ref:`deployment
   guides <ref_guide_deployment>`.

   Under Debian/Ubuntu, when adding the EdgeDB package repository, use this
   command instead:

   .. code-block:: bash

       $ echo deb [signed-by=/usr/local/share/keyrings/edgedb-keyring.gpg] \
           https://packages.edgedb.com/apt \
           $(grep "VERSION_CODENAME=" /etc/os-release | cut -d= -f2) main \
           | sudo tee /etc/apt/sources.list.d/edgedb.list

   Use this command for installation under Debian/Ubuntu:

   .. code-block:: bash

>        $ sudo apt-get update && sudo apt-get install edgedb-5

Author: I'm not sure, but I think it would all be under `edgedb-5`, although I haven't used that particular way of installing it on my Ubuntu PC.

   branch. Our CLI tools may need to first clone the "main" branch with the
   data into a "temp" branch. Then we can introspect the migration histories
   of "temp" and "feature" branches so that we can establish where they
   diverge. Take all the divergent migrations from the "feature" branch and
   apply them to the "temp" branch. If the operation is successful, drop the
   "feature" branch and rename "temp" to "feature". We now have successfully
   rebased "feature" branch on top of "main".

7) Since the state of "feature" is now a straightforward fast-forward w.r.t.
   the "main" branch we can finally merge "feature" back into main in VCS and
   then merge the EdgeDB branch as well (or rename "feature" EdgeDB branch
   into "main", if the old branch is no longer needed).

Update pgvector extension
-------------------------

A new HNSW (Hierarchical Navigable Small Worlds) index has been added to the ``pgvector`` extension. Just like IVFFlat indexes there are three flavors of HNSW corresponding to different operations:

* ``ext::pgvector::hnsw_euclidean``
* ``ext::pgvector::hnsw_ip``
* ``ext::pgvector::hnsw_cosine``

> We have also updated the mechanism for tuning all of the indexes provided in this extension. The ``probes`` (for IVFFlat) and ``ef_search`` (for HNSW) parameters can now be accessed via ``ext::pgvector::Config`` object.

Reviewer 2: ```suggestion
We have also updated the mechanism for tuning all of the indexes provided in this extension. The ``probes`` (for IVFFlat) and ``ef_search`` (for HNSW) parameters can now be accessed via the ``ext::pgvector::Config`` object.
```

The current config value can be found using this query:

.. code-block:: edgeql-repl

    db> select ext::pgvector::Config.probes;
    {1}

> Updating the value can be done by ``configure session`` command:

Reviewer 2: ```suggestion
Updating the value can be done using the ``configure session`` command:
```

.. code-block:: edgeql-repl

    db> configure session
    ... set ext::pgvector::Config::probes := 5;
    OK: CONFIGURE SESSION

It is also possible to restore the default config value:

.. code-block:: edgeql-repl

    db> configure session reset ext::pgvector::Config::probes;
    OK: CONFIGURE SESSION

Authentication
--------------

> We've expanded our ``auth`` extension with "WebAuthn" and "Magic Links" functionality.

Reviewer 2: Hey, @scotttrinh. Is there more we should say here?

Reviewer 3: Yep.

```suggestion
We're bringing two popular "passwordless" authentication schemes to our ``auth`` extension: the Web Authentication API (commonly known as WebAuthn or Passkeys), as well as email-based "magic links".

We've also added two popular chat platforms to our list of supported OAuth providers: Slack and Discord.
```

---

We're also missing these two PRs:

- https://github.com/edgedb/edgedb/pull/6708
- https://github.com/edgedb/edgedb/pull/6720

Author: I was thinking that while I had only the brief details in this section I would include PRs (so that there's some indicator of what goes there), but typically we only include PRs in the "Additional Changes" where most of them are mentioned rather tersely. In these "dedicated" sections we tend to just have text descriptions and examples rather than PR links.

* Add WebAuthn authentication.
  (:eql:gh:`#6791`, :eql:gh:`#6792`)

* Add Magic Links passwordless provider.
  (:eql:gh:`#6826`)

* Update auth UI to support multiple email factor providers.
  (:eql:gh:`#6874`)

* Move app details configuration into ``AuthConfig``.
  (:eql:gh:`#6754`)

Additional changes
==================

Performance
-----------

> We've made some changes to query caching aimed at improving the query recompilation in case of schema changes.

Reviewer 2: ```suggestion
We've made some changes to query caching aimed at improving query recompilation in case of schema changes.
```

We've also improved processing of large schemas and migrations.

EdgeQL
------

* Allow omitting ``union`` in ``for`` if the body is a statement.
  (:eql:gh:`#6810`)

>   If the ``for`` query body involves a statement such as ``insert``, ``update``, ``delete``, etc. you no longer need to write the ``union`` keyword and parenthesise the statement expression:

Reviewer 2: ```suggestion
  If the ``for`` query body involves a statement such as ``insert``, ``update``, ``delete``, etc., you no longer need to write the ``union`` keyword and add parentheses around the statement expression:
```

  .. code-block:: edgeql-diff

      for name int {'Alice', 'Billie', 'Cameron'}
    - union (
      insert User { name := name }
    - )

* Add ``administer vacuum()`` command.
  (:eql:gh:`#6663`)

>   The command ``administer vacuum()`` can take a list of object types, multi properties, multi links or links with link properties. There is also named only argument ``full`` that controls how thorough the cleaning is going to be. All of the arguments can be omitted. In case no target types are specified everything accessible to the user will be vacuumed.

Reviewer 2: ```suggestion
  The command ``administer vacuum()`` can take a list of object types, multi properties, multi links or links with link properties. There is also a named only argument ``full`` that controls how thorough the cleaning will be. All of the arguments can be omitted. In case no target types are specified, everything accessible to the user will be vacuumed.
```

config properties you need to use the type filter :eql:op:`[is
ext::pgvector::Config] <isintersect>`:

.. code-block:: edgeql-repl

    db> select cfg::Config.extensions[is ext::pgvector::Config]{*};
    {
      ext::pgvector::Config {
        id: 12b5c70f-0bb8-508a-845f-ca3d41103b6f,
        probes: 1,
        ef_search: 40,
      },
    }

Updating the value can be done using the ``configure session`` command:

.. code-block:: edgeql-repl

    db> configure session
    ... set ext::pgvector::Config::probes := 5;
    OK: CONFIGURE SESSION

It is also possible to restore the default config value:

.. code-block:: edgeql-repl

    db> configure session reset ext::pgvector::Config::probes;
    OK: CONFIGURE SESSION

Authentication
--------------

We're bringing two popular "passwordless" authentication schemes to our
``auth`` extension: the Web Authentication API (commonly known as WebAuthn or
Passkeys), as well as email-based "magic links".

We've also added two popular chat platforms to our list of supported OAuth
providers: Slack and Discord.

Additional changes
==================

Performance
-----------

> We've made some changes to query caching aimed at improving query

Reviewer 4: > The query compilation cache is now persisted across restarts, and cached queries are automatically recompiled after migrations are applied.
